src/main/java/org/springframework/data/aot/TypeCollector.java

```diff
@@ -0,0 +1,243 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.aot;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.springframework.core.ResolvableType;
+import org.springframework.data.util.Lazy;
+import org.springframework.util.ObjectUtils;
+import org.springframework.util.ReflectionUtils;
+
+/**
+ * @author Christoph Strobl
+ * @author Sebastien Deleuze
+ */
+public class TypeCollector {
+
+	private static Log logger = LogFactory.getLog(TypeCollector.class);
+
+	static final Set<String> EXCLUDED_DOMAINS = new HashSet<>(Arrays.asList("java", "sun.", "jdk.", "reactor.",
+			"kotlinx.", "kotlin.", "org.springframework.core.", "org.springframework.boot."));
+
+	private Predicate<Class<?>> excludedDomainsFilter = (type) -> {
+		return EXCLUDED_DOMAINS.stream().noneMatch(type.getPackageName()::startsWith);
+	};
+
+	Predicate<Class<?>> typeFilter = excludedDomainsFilter;
+
+	private final Predicate<Method> methodFilter = (method) -> {
+		if (method.getName().startsWith("$$_hibernate")) {
+			return false;
+		}
+		if (method.getDeclaringClass().getPackageName().startsWith("java.") || method.getDeclaringClass().isEnum()
+				|| EXCLUDED_DOMAINS.stream().anyMatch(it -> method.getDeclaringClass().getPackageName().startsWith(it))) {
+			return false;
+		}
+		if (method.isBridge() || method.isSynthetic()) {
+			return false;
+		}
+		return (!Modifier.isNative(method.getModifiers()) && !Modifier.isPrivate(method.getModifiers())
+				&& !Modifier.isProtected(method.getModifiers())) || !method.getDeclaringClass().equals(Object.class);
+	};
+
+	private Predicate<Field> fieldFilter = (field) -> {
+		if (field.isSynthetic() | field.getName().startsWith("$$_hibernate")) {
+			return false;
+		}
+		if (field.getDeclaringClass().getPackageName().startsWith("java.")) {
+			return false;
+		}
+		return true;
+	};
+
+	public TypeCollector filterFields(Predicate<Field> filter) {
+		this.fieldFilter = filter.and(filter);
```

Comment:
Just curious if you meant to prepend here?  I ask because the filterTypes(:Predicate) method below "appends".